File: parser/src/main/java/com/aaroncoplan/waterfall/parser/FileUtils.java
package com.aaroncoplan.waterfall.parser;

import java.io.File;
import java.io.FileNotFoundException;
import java.util.Scanner;

public class FileUtils {

    public static Pair<Boolean, String> isReadableFile(final String filePath) {
        final File file = new File(filePath);
        if (!file.exists()) {
            return new Pair<>(false, String.format("File '%s' does not exist", filePath));
        }
        if (!file.canRead()) {
            return new Pair<>(false, String.format("File '%s' cannot be read", filePath));
        }
        if (!file.isFile()) {
            return new Pair<>(false, String.format("File '%s' is not a file", filePath));
        }

        return new Pair<>(true, null);
    }

    static String readFile(final String filePath) {
        final StringBuilder contents = new StringBuilder();
        try {
            Scanner scanner = new Scanner(new File(filePath));
            while (scanner.hasNextLine()) {
                contents.append(scanner.nextLine());
                contents.append('\n');
            }
            scanner.close();
        } catch(FileNotFoundException e) {
            return null;
        }

        return contents.toString();
    }
}

File: parser/src/main/kotlin/com/aaroncoplan/waterfall/parser/FileUtils.kt
package com.aaroncoplan.waterfall.parser

import java.io.File
import java.io.FileNotFoundException
import java.util.Scanner

object FileUtils {

    @JvmStatic
    fun isReadableFile(filePath: String): Pair<Boolean, String?> {
        val file = File(filePath)
        if (!file.exists()) {
            return Pair(false, "File '$filePath' does not exist")
        }
        if (!file.canRead()) {
            return Pair(false, "File '$filePath' cannot be read")
        }
        if (!file.isFile) {
            return Pair(false, "File '$filePath' is not a file")
        }
        return Pair(true, null)
    }

    /** Package-private in the Java version; kept internal here for FileParser only. */
    @JvmStatic
    internal fun readFile(filePath: String): String? {
        val contents = StringBuilder()
        return try {
            Scanner(File(filePath)).use { scanner ->
                while (scanner.hasNextLine()) {
                    contents.append(scanner.nextLine())
                    contents.append('\n')
                }
            }
            contents.toString()
        } catch (e: FileNotFoundException) {
            null
        }
    }
}
